plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -1753,20 +1763,27 @@ else if (partitionUpdate.getUpdateMode() == APPEND) {
                         partitionUpdate.getFileNames(),
                         partitionStatistics);
             }
-            else if (partitionUpdate.getUpdateMode() == NEW || partitionUpdate.getUpdateMode() == OVERWRITE) {
+            else if (partitionUpdate.getUpdateMode() == NEW || partitionUpdate.getUpdateMode() == PartitionUpdate.UpdateMode.OVERWRITE) {
                 // insert into new partition or overwrite existing partition
                 Partition partition = buildPartitionObject(session, table, partitionUpdate);
                 if (!partition.getStorage().getStorageFormat().getInputFormat().equals(handle.getPartitionStorageFormat().getInputFormat()) && isRespectTableFormat(session)) {
                     throw new TrinoException(HIVE_CONCURRENT_MODIFICATION_DETECTED, "Partition format changed during insert");
                 }
-                if (partitionUpdate.getUpdateMode() == OVERWRITE) {
-                    metastore.dropPartition(session, handle.getSchemaName(), handle.getTableName(), partition.getValues(), true);
-                }
+                boolean directToExistingDir = handle.getLocationHandle().getWriteMode() == DIRECT_TO_TARGET_EXISTING_DIRECTORY;
                 PartitionStatistics partitionStatistics = createPartitionStatistics(
                         partitionUpdate.getStatistics(),
                         columnTypes,
                         getColumnStatistics(partitionComputedStatistics, partition.getValues()));
-                metastore.addPartition(session, handle.getSchemaName(), handle.getTableName(), partition, partitionUpdate.getWritePath(), partitionStatistics);
+                if (partitionUpdate.getUpdateMode() == PartitionUpdate.UpdateMode.OVERWRITE && directToExistingDir) {
+                    removeNonCurrentQueryFiles(session, partitionUpdate.getTargetPath());
+                }
+                else if (partitionUpdate.getUpdateMode() == PartitionUpdate.UpdateMode.OVERWRITE) {
+                    metastore.dropPartition(session, handle.getSchemaName(), handle.getTableName(), partition.getValues(), true);
+                    metastore.addPartition(session, handle.getSchemaName(), handle.getTableName(), partition, partitionUpdate.getWritePath(), partitionStatistics);
+                }
```

Comment:
Nit: I'd probably format this as

```
if (OVERWRITE) {
  if (directToExistingDir) {
    ...
  }
  else {
    ...
  }
}
```

Reply:
Heh I was having both, but ye let's keep int in less pretty but clearer way.